File: ios/Jazzify/Audio/YinPitch.swift
import Foundation

enum YinPitch {
    static let windowSize = 4096
    static let hopSize = 480
    static let sampleRate = 48_000.0
    static let frameSec = Double(hopSize) / sampleRate
    static let threshold = 0.15

    struct FrequencyRange {
        let minHz: Double
        let maxHz: Double
    }

    struct AnalysisResult {
        let frequencyHz: Double?
        let confidence: Double
        let volume: Double
    }

    static func frequencyRange(for shift: VoiceLowPitchShift) -> FrequencyRange {
        switch shift {
        case .off:
            return FrequencyRange(minHz: 80, maxHz: 700)
        case .plus12:
            return FrequencyRange(minHz: 55, maxHz: 700)
        case .plus24:
            return FrequencyRange(minHz: 30, maxHz: 400)
        }
    }

    static func usesPitchDetection(for shift: VoiceLowPitchShift) -> Bool {
        shift != .off
    }

    static func hzToMidi(_ frequencyHz: Double) -> Double {
        69 + 12 * log2(frequencyHz / 440)
    }

    static func scaleOnsetConfig(_ config: PitchOnsetTrackerConfig) -> PitchOnsetTrackerConfig {
        func halve(_ value: Int) -> Int { max(1, Int((Double(value) / 2).rounded())) }
        var scaled = config
        scaled.pitchStableFrames = halve(config.pitchStableFrames)
        scaled.releaseFrames = halve(config.releaseFrames)
        scaled.minNoteFrames = halve(config.minNoteFrames)
        scaled.retriggerGuardFrames = halve(config.retriggerGuardFrames)
        scaled.retriggerLookbackFrames = halve(config.retriggerLookbackFrames)
        return scaled
    }

    static func analyzeWindow(
        _ window: UnsafePointer<Float>,
        count: Int,
        sampleRate: Double,
        range: FrequencyRange,
        threshold: Double = YinPitch.threshold
    ) -> AnalysisResult {
        guard count > 0 else {
            return AnalysisResult(frequencyHz: nil, confidence: 0, volume: 0)
        }

        var sumSquares = 0.0
        for index in 0..<count {
            let sample = Double(window[index])
            sumSquares += sample * sample
        }
        let volume = sqrt(sumSquares / Double(count))
        guard volume >= 1e-6 else {
            return AnalysisResult(frequencyHz: nil, confidence: 0, volume: volume)
        }

        let minPeriod = max(2, Int(floor(sampleRate / range.maxHz)))
        let maxPeriod = min(count - 1, Int(ceil(sampleRate / range.minHz)))
        guard minPeriod < maxPeriod else {
            return AnalysisResult(frequencyHz: nil, confidence: 0, volume: volume)
        }

        var difference = [Double](repeating: 0, count: maxPeriod + 1)
        for tau in 1...maxPeriod {
            var sum = 0.0
            let limit = count - tau
            for index in 0..<limit {
                let delta = Double(window[index]) - Double(window[index + tau])
                sum += delta * delta
            }
            difference[tau] = sum
        }

        var cmndf = [Double](repeating: 1, count: maxPeriod + 1)
        cmndf[0] = 1
        var runningSum = 0.0
        for tau in 1...maxPeriod {
            runningSum += difference[tau]
            cmndf[tau] = runningSum > 0 ? (difference[tau] * Double(tau)) / runningSum : 1
        }

        var bestTau = -1
        var bestValue = 1.0
        var tau = minPeriod
        while tau <= maxPeriod {
            let value = cmndf[tau]
            if value < threshold {
                var localTau = tau
                while localTau + 1 <= maxPeriod, cmndf[localTau + 1] < value {
                    localTau += 1
                }
                bestTau = localTau
                bestValue = cmndf[localTau]
                break
            }
            tau += 1
        }

        if bestTau < 0 {
            for candidate in minPeriod...maxPeriod {
                let value = cmndf[candidate]
                if value < bestValue {
                    bestValue = value
                    bestTau = candidate
                }
            }
        }

        guard bestTau > 0, bestValue < 0.4 else {
            return AnalysisResult(frequencyHz: nil, confidence: 0, volume: volume)
        }

        let refinedTau = parabolicInterpolation(cmndf: cmndf, tau: bestTau)
        let frequencyHz = sampleRate / refinedTau
        guard frequencyHz.isFinite, frequencyHz >= range.minHz, frequencyHz <= range.maxHz else {
            return AnalysisResult(frequencyHz: nil, confidence: 0, volume: volume)
        }

        let confidence = max(0, min(1, 1 - bestValue))
        return AnalysisResult(frequencyHz: frequencyHz, confidence: confidence, volume: volume)
    }

    private static func parabolicInterpolation(cmndf: [Double], tau: Int) -> Double {
        guard tau > 0, tau < cmndf.count - 1 else { return Double(tau) }
        let s0 = cmndf[tau - 1]
        let s1 = cmndf[tau]
        let s2 = cmndf[tau + 1]
        let denominator = 2 * s1 - s2 - s0
        guard denominator != 0 else { return Double(tau) }
        return Double(tau) + (s2 - s0) / (2 * denominator)
    }
}

/// 推論スレッド専用。4096 サンプル窓を 10 ms ごとに YIN 解析する。
final class YinPitchProcessor {
    private var shift = VoiceLowPitchShift.off
    private var ring = [Float](repeating: 0, count: YinPitch.windowSize)
    private var ringWriteIndex = 0
    private var filledSamples = 0
    private var samplesSinceHop = 0
    private var orderedWindow = [Float](repeating: 0, count: YinPitch.windowSize)

    func reset() {
        for index in ring.indices {
            ring[index] = 0
        }
        ringWriteIndex = 0
        filledSamples = 0
        samplesSinceHop = 0
    }

    func setShift(_ shift: VoiceLowPitchShift) {
        guard shift != self.shift else { return }
        self.shift = shift
        reset()
    }

    func push(
        samples: UnsafePointer<Float>,
        count: Int,
        emit: (_ result: YinPitch.AnalysisResult) -> Void
    ) {
        let range = YinPitch.frequencyRange(for: shift)
        for index in 0..<count {
            ring[ringWriteIndex] = samples[index]
            ringWriteIndex = (ringWriteIndex + 1) % YinPitch.windowSize
            if filledSamples < YinPitch.windowSize {
                filledSamples += 1
            }

            samplesSinceHop += 1
            guard samplesSinceHop >= YinPitch.hopSize, filledSamples >= YinPitch.windowSize else { continue }
            samplesSinceHop = 0

            var readIndex = ringWriteIndex
            for offset in 0..<YinPitch.windowSize {
                orderedWindow[offset] = ring[readIndex]
                readIndex = (readIndex + 1) % YinPitch.windowSize
            }

            orderedWindow.withUnsafeBufferPointer { buffer in
                guard let base = buffer.baseAddress else { return }
                let result = YinPitch.analyzeWindow(
                    base,
                    count: YinPitch.windowSize,
                    sampleRate: YinPitch.sampleRate,
                    range: range
                )
                emit(result)
            }
        }
    }
}

File: src/utils/pitchInput/yinPitch.ts
/**
 * yinPitch - 低音向け YIN（累積平均正規化差分）F0 推定。
 * Web / iOS で同一係数・同一探索帯域を使う。
 */

import type { PitchOnsetTrackerConfig } from '@/utils/pitchInput/pitchOnsetTracker';
import {
  PITCH_SAMPLE_RATE,
  type VoiceLowPitchShift,
} from '@/utils/pitchInput/pitchOctaveUpsample';

export const YIN_WINDOW_SIZE = 4096;
export const YIN_HOP_SIZE = 480;
export const YIN_FRAME_SEC = YIN_HOP_SIZE / PITCH_SAMPLE_RATE;
export const YIN_THRESHOLD = 0.15;

export interface YinFrequencyRange {
  readonly minHz: number;
  readonly maxHz: number;
}

export interface YinAnalysisResult {
  readonly frequencyHz: number | null;
  readonly confidence: number;
  readonly volume: number;
}

export const yinFrequencyRangeForShift = (shift: VoiceLowPitchShift): YinFrequencyRange => {
  if (shift === 12) {
    return { minHz: 55, maxHz: 700 };
  }
  if (shift === 24) {
    return { minHz: 30, maxHz: 400 };
  }
  return { minHz: 80, maxHz: 700 };
};

export const usesYinPitchDetection = (shift: VoiceLowPitchShift): boolean => shift !== 0;

export const hzToMidi = (frequencyHz: number): number =>
  69 + (12 * Math.log2(frequencyHz / 440));

export const scaleOnsetConfigForYin = (
  config: PitchOnsetTrackerConfig,
): PitchOnsetTrackerConfig => {
  const halve = (value: number): number => Math.max(1, Math.round(value / 2));
  return {
    ...config,
    pitchStableFrames: halve(config.pitchStableFrames),
    releaseFrames: halve(config.releaseFrames),
    minNoteFrames: halve(config.minNoteFrames),
    retriggerGuardFrames: halve(config.retriggerGuardFrames),
    retriggerLookbackFrames: halve(config.retriggerLookbackFrames),
  };
};

const computeRms = (window: Float32Array): number => {
  let sum = 0;
  for (let index = 0; index < window.length; index += 1) {
    const sample = window[index] ?? 0;
    sum += sample * sample;
  }
  return Math.sqrt(sum / window.length);
};

const parabolicInterpolation = (
  cmndf: Float64Array,
  tau: number,
): number => {
  if (tau <= 0 || tau >= cmndf.length - 1) {
    return tau;
  }
  const s0 = cmndf[tau - 1] ?? 0;
  const s1 = cmndf[tau] ?? 0;
  const s2 = cmndf[tau + 1] ?? 0;
  const denominator = 2 * s1 - s2 - s0;
  if (denominator === 0) {
    return tau;
  }
  return tau + (s2 - s0) / (2 * denominator);
};

export const analyzeYinWindow = (
  window: Float32Array,
  sampleRate: number,
  range: YinFrequencyRange,
  threshold = YIN_THRESHOLD,
): YinAnalysisResult => {
  const volume = computeRms(window);
  if (volume < 1e-6) {
    return { frequencyHz: null, confidence: 0, volume };
  }

  const minPeriod = Math.max(2, Math.floor(sampleRate / range.maxHz));
  const maxPeriod = Math.min(
    window.length - 1,
    Math.ceil(sampleRate / range.minHz),
  );
  if (minPeriod >= maxPeriod) {
    return { frequencyHz: null, confidence: 0, volume };
  }

  const difference = new Float64Array(maxPeriod + 1);
  for (let tau = 1; tau <= maxPeriod; tau += 1) {
    let sum = 0;
    for (let index = 0; index < window.length - tau; index += 1) {
      const delta = (window[index] ?? 0) - (window[index + tau] ?? 0);
      sum += delta * delta;
    }
    difference[tau] = sum;
  }

  const cmndf = new Float64Array(maxPeriod + 1);
  cmndf[0] = 1;
  let runningSum = 0;
  for (let tau = 1; tau <= maxPeriod; tau += 1) {
    runningSum += difference[tau] ?? 0;
    cmndf[tau] = runningSum > 0 ? ((difference[tau] ?? 0) * tau) / runningSum : 1;
  }

  let bestTau = -1;
  let bestValue = 1;
  for (let tau = minPeriod; tau <= maxPeriod; tau += 1) {
    const value = cmndf[tau] ?? 1;
    if (value >= threshold) {
      continue;
    }
    while (tau + 1 <= maxPeriod && (cmndf[tau + 1] ?? 1) < value) {
      tau += 1;
    }
    bestTau = tau;
    bestValue = cmndf[tau] ?? 1;
    break;
  }

  if (bestTau < 0) {
    for (let tau = minPeriod; tau <= maxPeriod; tau += 1) {
      const value = cmndf[tau] ?? 1;
      if (value < bestValue) {
        bestValue = value;
        bestTau = tau;
      }
    }
  }

  if (bestTau <= 0 || bestValue >= 0.4) {
    return { frequencyHz: null, confidence: 0, volume };
  }

  const refinedTau = parabolicInterpolation(cmndf, bestTau);
  const frequencyHz = sampleRate / refinedTau;
  if (!Number.isFinite(frequencyHz) || frequencyHz < range.minHz || frequencyHz > range.maxHz) {
    return { frequencyHz: null, confidence: 0, volume };
  }

  const confidence = Math.max(0, Math.min(1, 1 - bestValue));
  return { frequencyHz, confidence, volume };
};

export class YinPitchProcessor {
  private readonly windowSize = YIN_WINDOW_SIZE;
  private readonly hopSize = YIN_HOP_SIZE;
  private readonly ring = new Float32Array(YIN_WINDOW_SIZE);
  private ringWriteIndex = 0;
  private filledSamples = 0;
  private samplesSinceHop = 0;
  private shift: VoiceLowPitchShift = 0;

  reset(): void {
    this.ring.fill(0);
    this.ringWriteIndex = 0;
    this.filledSamples = 0;
    this.samplesSinceHop = 0;
  }

  setShift(shift: VoiceLowPitchShift): void {
    if (shift === this.shift) {
      return;
    }
    this.shift = shift;
    this.reset();
  }

  push(samples: Float32Array): YinAnalysisResult[] {
    const results: YinAnalysisResult[] = [];
    const range = yinFrequencyRangeForShift(this.shift);

    for (let index = 0; index < samples.length; index += 1) {
      this.ring[this.ringWriteIndex] = samples[index] ?? 0;
      this.ringWriteIndex = (this.ringWriteIndex + 1) % this.windowSize;
      if (this.filledSamples < this.windowSize) {
        this.filledSamples += 1;
      }

      this.samplesSinceHop += 1;
      if (this.samplesSinceHop < this.hopSize || this.filledSamples < this.windowSize) {
        continue;
      }
      this.samplesSinceHop = 0;

      const window = new Float32Array(this.windowSize);
      let readIndex = this.ringWriteIndex;
      for (let offset = 0; offset < this.windowSize; offset += 1) {
        window[offset] = this.ring[readIndex] ?? 0;
        readIndex = (readIndex + 1) % this.windowSize;
      }
      results.push(analyzeYinWindow(window, PITCH_SAMPLE_RATE, range));
    }

    return results;
  }
}
